proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -3310,14 +3310,14 @@ lemma possible_switch_to_valid_sched_strong:
   by (auto simp: obj_at_kh_kheap_simps valid_sched_valid_sched_except_blocked vs_all_heap_simps
                  in_cur_domain_def etcb_at_def valid_sched_def not_cur_thread_def ct_in_cur_domain_def)
 
-crunch etcb_at[wp]: awaken "etcb_at P t"
```

Comment:
I wonder if we could make `awaken`/`whileLoop` crunchable by instantiating the `whileLoop_wp` rule with one invariant predicate for all parts such that it works without further instantiation (and then giving that rule to crunch as `wp`).

Have you tried something like that?

Reply:
Agreed, that's what I was thinking as well. Something like the following seems like it should work for this.
```
lemma whileLoop_wp':
  "(⋀r. ⦃ λs. I r s ∧ C r s ⦄ B r ⦃ I ⦄) ⟹
  ⦃ I r ⦄ whileLoop C B r ⦃ I ⦄"